examples/htool.c

```diff
@@ -251,6 +251,413 @@ static int do_target_reset(uint32_t reset_option) {
+  }
+  return max_size;
+}
+
+size_t size_to_send(const uint8_t* data, size_t size) {
+  for (int i = size; i > 0; i--) {
+    if (data[i - 1] != 0xff) return i;
+  }
+  return 0;
+}
+
+struct payload_update_packet {
+  uint32_t offset; /* image offset */
+  uint32_t len;    /* packet length excluding this header */
+  uint8_t type;    /* One of PAYLOAD_UPDATE_* */
+  /* payload data immediately follows */
+} __attribute__((packed));
+
+static int send_payload_update_command(struct libhoth_usb_device* dev,
+                                       uint8_t command, uint32_t offset,
+                                       uint32_t len) {
+  struct payload_update_packet request;
+  request.offset = offset;
+  request.len = len;
+  request.type = command;
+
+  uint8_t response[1];  // ignored
+  size_t rlen = 0;
+
+  int ret = htool_exec_hostcmd(
+      dev, EC_CMD_BOARD_SPECIFIC_BASE + EC_PRV_CMD_HOTH_PAYLOAD_UPDATE, 0,
+      &request, sizeof(request), &response, sizeof(response), &rlen);
+  return ret;
+}
+
+static int send_payload_update_command_with_return_value(struct libhoth_usb_device* dev,
+                                       uint8_t command, uint32_t offset,
+                                       uint32_t len, uint8_t* buf) {
+  struct payload_update_packet request;
+  request.offset = offset;
+  request.len = len;
+  request.type = command;
+  size_t rlen = 0;
+
+  int ret = htool_exec_hostcmd(
+      dev, EC_CMD_BOARD_SPECIFIC_BASE + EC_PRV_CMD_HOTH_PAYLOAD_UPDATE, 0,
+      &request, sizeof(request), buf, len, &rlen);
+  return ret;
+}
+
+int command_payload_read(struct libhoth_usb_device* dev, uint32_t offset, uint32_t len, uint8_t* buf) {
+  return send_payload_update_command_with_return_value(dev, 6, offset, len, buf);
+}
+
+// Similar to Hoth::getPayloadSize
+static int command_payload_size(const struct htool_invocation* inv) {
+  printf("The payload_size program uses binary-search to determine the payload size.\n"
+         "Error message are expected. Just disregard the errors.\n");
+  struct libhoth_usb_device* dev = htool_libhoth_usb_device();
+  if (!dev) {
+    return -1;
+  }
+  uint32_t right = 1;
+  uint8_t buf[1];
+  while (true) {
+    int ret = command_payload_read(dev, right, 1, buf);
+    if (ret != 0) {
+      break;
+    } else {
+      right *= 2;
+    }
+  }
+
+  uint32_t left = right / 2;
+  if (right > 1) {
+    if (command_payload_read(dev, right-1, 1, buf) == 0) {
+      left = right;
+      goto DONE;
+    }
+  }
+
+  while (left < right) {
+    uint32_t mid = (left+right) / 2;
+    if (command_payload_read(dev, mid, 1, buf) != 0) {
+      right = mid;
+    } else {
+      left = mid + 1;
+    }
+  }
+
+  DONE:
+  printf("Payload size is %u\n", right);
+}
+
+static int command_payload_update(const struct htool_invocation* inv) {
+  struct libhoth_usb_device* dev = htool_libhoth_usb_device();
+  if (!dev) {
+    return -1;
+  }
+
+  bool dry_run = false;
+  char* x = getenv("DRY_RUN");
```

Comment:
Having dry-run be an environment variable is a little weird -- I think this would more naturally be a command-line flag? That might be inconvenient here, though.

Reply:
It is a debugging statement, and will be removed.